plugins/modules/packaging/os/pacman_key.py

```diff
@@ -0,0 +1,334 @@
+
+- name: Import a key via url
+  pacman_key:
+    id: 01234567890ABCDE01234567890ABCDE12345678
+    url: https://domain.tld/keys/keyfile.asc
+    state: present
+
+- name: Import a key via keyserver
+  pacman_key:
+    id: 01234567890ABCDE01234567890ABCDE12345678
+    keyserver: keyserver.domain.tld
+
+- name: Import a key into an alternative keyring
+  pacman_key:
+    id: 01234567890ABCDE01234567890ABCDE12345678
+    file: /tmp/keyfile.asc
+    keyring: /etc/pacman.d/gnupg-alternative
+
+- name: Remove a key from the keyring
+  pacman_key:
+    id: 01234567890ABCDE01234567890ABCDE12345678
+    state: absent
+'''
+
+RETURN = r''' # '''
+
+import os.path
+import tempfile
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.urls import fetch_url
+from ansible.module_utils._text import to_native
+
+
+class PacmanKey(object):
+    def __init__(self, module):
+        self.module = module
+        self.gpg = module.get_bin_path('gpg', required=True)
+        self.pacman_key = module.get_bin_path('pacman-key', required=True)
+        keyid = module.params['id']
+        url = module.params['url']
+        data = module.params['data']
+        file = module.params['file']
+        keyserver = module.params['keyserver']
+        fingerprint = module.params['fingerprint']
+        force_update = module.params['force_update']
+        keyring = module.params['keyring']
+        state = module.params['state']
+
+        keyid = self.sanitise_keyid(keyid)
+        if fingerprint:
+            fingerprint = self.sanitise_fingerprint(fingerprint)
+        key_present = self.key_in_keyring(keyid, keyring)
+
+        if module.check_mode:
+            if state == "present":
+                if (key_present and force_update) or not key_present:
+                    module.exit_json(changed=True)
+                module.exit_json(changed=False)
+            elif state == "absent":
+                if key_present:
+                    module.exit_json(changed=True)
+                module.exit_json(changed=False)
+
+        if state == "present":
+            if key_present and not force_update:
+                module.exit_json(changed=False)
+
+            if data:
+                file = self.save_key(data)
+                self.add_key(file, keyid, fingerprint, keyring)
+                module.exit_json(changed=True)
+            elif file:
+                self.add_key(file, keyid, fingerprint, keyring)
+                module.exit_json(changed=True)
+            elif url:
+                data = self.fetch_key(url)
+                file = self.save_key(data)
+                self.add_key(file, keyid, fingerprint, keyring)
+                module.exit_json(changed=True)
+            elif keyserver:
+                self.recv_key(keyid, keyserver, keyring)
+                module.exit_json(changed=True)
+            else:
+                module.fail_json(msg="expected one of: data, file, url, keyserver. got none")
```

Comment:
Check this earlier. check_mode should also fail when this is not the case, and the module should also fail if this is not the case and they key already exists in the keyring.

Reply:
Latest commit _should_ fix this. It's a bit early in the morning to be coding, do you mind checking this over?

Reply:
Looks good to me!

Reply:
Fantastic. I’ll throw together some unit tests. Can you recommend any examples in pre-existing modules or documentation that I should browse?

Reply:
Not in particular. The tests for the ufw module look helpful: https://github.com/ansible-collections/community.general/blob/main/tests/unit/plugins/modules/system/test_ufw.py The ufw module is mainly executing the `ufw` binary and reacts on its output, so you can get some ideas from there.